import operator
import os

SNAKEMAKE_DIR = os.path.dirname(workflow.snakefile)

# If the user has a config file in the current working directory, use
# that. Otherwise, use SMRT SV defaults.
if os.path.exists("config.json"):
    configfile: "config.json"
else:
    configfile: "%s/../config.template.json" % SNAKEMAKE_DIR

INPUT_READS = config.get("reads", "input.fofn")
CELERA_SPEC = config.get("celera_spec", "%s/../celera/pacbio.local_human.spec" % SNAKEMAKE_DIR)
CHROMOSOME_LENGTHS = config.get("reference_index", "%s.fai" % config["reference"])
print("#######"+CHROMOSOME_LENGTHS)

# Get a sample .bax.h5 file from the given list of input reads.
#with open(INPUT_READS, "r") as fh:
#    BAS_TEMPLATE = next(fh).strip()

# By default delay jobs for a random amount of time up to this value to prevent
# too much I/O from simultaneous assemblies.
DEFAULT_MAX_DELAY = 1

# User-defined file of alignments with one absolute path to a BAM per line.
ALIGNMENTS = config["alignments"]
print("******"+ALIGNMENTS)

# User-defined region in form of "{chrom}-{start}-{end}"
REGION = config["region"]

# Convert filesystem-safe filename of "chrom-start-end" to the more-standard
# region of "chrom:start-end"
STANDARD_REGION = REGION.replace("-", ":", 1)

# Calculate the size of the given region of format "chrom-start-end" by
# splitting the region on "-", selecting the last two elements, reversing their
# order to produce (end, start), converting strings to integers, and applying
# the subtraction operator.
REGION_SIZE = str(operator.sub(*map(int, reversed(REGION.split("-")[1:3]))))

#
# Define rules.
#

#use13
rule align_consensus_to_reference_region:
    input: "consensus.trimmed.fasta", "reference_region.fasta"
    output: "consensus_reference_alignment.sam"
    params: sge_opts="", mapping_quality_threshold=str(config["mapping_quality"]), alignment_parameters=config["alignment_parameters"]
    shell: """blasr {input} -clipping subread -out /dev/stdout -sam {params.alignment_parameters} | samtools view -q {params.mapping_quality_threshold} - | awk 'OFS="\\t" {{ sub(/:/, "-", $3); num_of_pieces=split($3, pieces, "-"); $3 = pieces[1]; $4 = pieces[2] + $4; print }}' | sed 's/RG:Z:\w\+\\t//' > {output};"""
        "echo 13_{params.alignment_parameters}_{params.mapping_quality_threshold}"

# use12
rule extract_reference_sequence:
    input: reference=config["reference"], region="reference_region.bed"
    output: temp("reference_region.fasta")
    params: sge_opts=""
    shell: "bedtools getfasta -fi {input.reference} -bed {input.region} -fo {output};"
        "echo 12_{input.reference}_{output}"

# use11
rule create_reference_region:
    output: "reference_region.bed"
    params: sge_opts=""
    shell: "echo {REGION} | sed 's/-/\\t/g' > {output};"
        "echo 11"

# use10
rule trim_consensus:
    input: "consensus.fasta"
    output: "consensus.trimmed.fasta"
    params: sge_opts=""
    shell: "{SNAKEMAKE_DIR}/../scripts/trim_lowercase.py {input} {output};"
        "echo 10"

# use9
rule quiver_assembly:
    input: assembly="assembly.fasta", alignments="alignment.cmp.h5", assembly_index="assembly.fasta.fai"
    output: "consensus.fasta"
    params: sge_opts="", threads="4"
    run:
        shell("""echo -e "{REGION}\tquiver_failed" >> %s""" % config["log"])
        shell("cat {input.assembly} > {output}")
        shell("echo 9")
#        try:
#            shell("quiver -j {params.threads} --referenceFilename {input.assembly} {input.alignments} -o {output}; ")
#            shell("sed -i 's/^>\(.\+\)/>{REGION}|\\1/' {output}; ")
#        except:
#            shell("""echo -e "{REGION}\tquiver_failed" >> %s""" % config["log"])
#            shell("cat {input.assembly} > {output}")

# use8
rule convert_assembly_alignments_to_hdf5:
#    input: alignments="alignment.sam", assembly="assembly.fasta", reads="reads.bas.h5"
    input: alignments="alignment.sam", assembly="assembly.fasta"
    output: "alignment.cmp.h5"
    params: sge_opts=""
    shell:
         "cp {input.alignments} {output};"
         "echo 8"

#        "samtoh5 {input.alignments} {input.assembly} {output} -useShortRefName; "
#        "cmph5tools.py sort --deep {output}; "
#        "loadPulses {input.reads} {output} -metrics InsertionQV,DeletionQV,SubstitutionQV,MergeQV,SubstitutionTag,DeletionTag; "
#        "{SNAKEMAKE_DIR}/../scripts/LoadGenericChemistry.py {BAS_TEMPLATE} {output}"

# use7
rule map_reads_to_assembly:
#    input: "reads.bas.h5", "assembly.fasta"
    input: "reads.fasta", "assembly.fasta"
    output: "alignment.sam"
    params: sge_opts="", threads="4"
    shell: "blasr {input} -sam -bestn 1 -out /dev/stdout -nproc {params.threads} | samtools view -h -F 0x4 -S - > {output};"
        "echo 7_{input}_{params.threads}"

# use6
rule index_assembly:
    input: "assembly.fasta"
    output: "assembly.fasta.fai"
    params: sge_opts=""
    shell: "samtools faidx {input};"
        "echo 6"

# use5
rule assemble_reads:
    input: "reads.fastq"
    output: "assembly.fasta"
    params: sge_opts="-l mfree=4G -pe serial 2 -l disk_free=10G", threads="4", read_length="1000", partitions="50", max_runtime="10m"
    run:
        assembly_output = "local/9-terminator/asm.ctg.fasta"
        unitig_output = "local/9-terminator/asm.utg.fasta"
        assembly_exists = False

        try:
            shell("timeout {params.max_runtime} PBcR -threads {params.threads} -length {params.read_length} -partitions {params.partitions} -l local -s {CELERA_SPEC} -fastq {input} genomeSize={REGION_SIZE} assembleMinCoverage=5 &> assembly.log")
            shell("echo 5_try_{params.max_runtime}_{params.threads}_{params.read_length}_{params.partitions}_{CELERA_SPEC}_{REGION_SIZE}")
            5_try_10m_4_1000_50_/share/home/chuanlex/xieshangqian/software/smrtsv/pacbio_variant_caller/rules/../celera/pacbio.local_human.spec_60000

        except:
            shell("""echo -e "{REGION}\tassembly_crashed" >> %s""" % config["log"])
            shell("echo 5_except")

        if os.path.exists(assembly_output) and os.stat(assembly_output).st_size > 0:
            shell("cat {assembly_output} > {output}")
            shell("""echo -e "{REGION}\tassembly_exists" >> %s""" % config["log"])
            assembly_exists = True
            shell("echo 5_if")
        elif os.path.exists(unitig_output) and os.stat(unitig_output).st_size > 0:
            shell("cat {unitig_output} > {output}")
            shell("""echo -e "{REGION}\tunitig_assembly_exists" >> %s""" % config["log"])
            assembly_exists = True
            shell("echo 5_elif")
        else:
            shell("""echo -e "{REGION}\tno_assembly_exists" >> %s""" % config["log"])
            shell("echo 5_else")

        # Create an empty assembly for failed regions.
        if not assembly_exists:
            shell("echo -e '>{REGION}\nN' > {output}")

#rule convert_reads_to_bas:
#    input: "reads.bam"
#    output: "reads.bas.h5"
#    params: sge_opts=""
#    shell: "samtools view -h {input} | samtobas /dev/stdin {output}"

# use4
rule convert_reads_to_fastq:
    input: "reads.fasta"
    output: "reads.fastq"
    params: sge_opts=""
    shell: "{SNAKEMAKE_DIR}/../scripts/FastaToFakeFastq.py {input} {output};"
        "echo 4"

# use3
rule convert_reads_to_fasta:
    input: "reads.bam"
    output: "reads.fasta"
    params: sge_opts=""
    shell: """samtools view {input} | awk '{{ print ">"$1; print $10 }}' | {SNAKEMAKE_DIR}/../scripts/FormatFasta.py --fakename > {output};"""
        "echo 3"

# use2
rule convert_reads_to_bam:
    input: "reads.sam"
    output: "reads.bam"
    params: sge_opts=""
    shell: "samtools view -buS {input} > {output};"
        "echo 2"

# use1
rule get_reads:
    input: ALIGNMENTS
    output: "reads.sam"
    params: sge_opts="", mapping_quality_threshold=str(config["mapping_quality"]), max_delay=str(config.get("minutes_to_delay_jobs", DEFAULT_MAX_DELAY))
    shell:
        "sleep $[ ( $RANDOM % {params.max_delay} ) ]m; "
        "head -n 1 {input} | xargs -i samtools view -H {{}} > {output}; "
        "cat {input} | xargs -i samtools view -q {params.mapping_quality_threshold} {{}} {STANDARD_REGION} >> {output};"
        "echo 1_{params.max_delay}_{input}_{params.mapping_quality_threshold}_{STANDARD_REGION}"
